### DISCUSSION TITLE
Garage Door Opener

### DISCUSSION THREAD

**Discussion by camerong3** [HomeSpan/HomeSpan#669] [Q&A]
I'm having some trouble getting my HomeKit Garage door project working. I have a relay controlling the garage door motor as well as two hall sensors to detect whether the door is completely closed or completely open. I've attached my code below. I want it to update the status of the HomeKit door depending on the status of the hall sensors. For example, if the closed sensor is near the magnet the door should display closed. For relay motor control, it should pulse the relay open then closed in order to trigger the garage door motor to operate. I would appreciate some help with the code. Thanks! A future goal of this project is to also have a window shade accessory where the garage door can be opened or closed incrementally with the hall sensors at the two extremes (max open and completely closed).
______________________________________
File "Garagedoor.ino"
______________________________________
#include "HomeSpan.h"
#include "DEV_DoorsWindows.h"

void setup() {
  Serial.begin(115200);
  homeSpan.setPairingCode("212-40-168");

  homeSpan.begin(Category::Bridges,"HomeSpan Bridge");

  new SpanAccessory();  
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      
  new SpanAccessory();                                                          
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      new Characteristic::Name("Garage Door");
    new DEV_GarageDoor();                                                               // Create a Garage Door Opener (see DEV_DoorsWindows.h for definition)                                                             // Create a motorized Window Shade (see DEV_DoorsWindows.h for definition)

  pinMode(closeSensorPIN, INPUT_PULLUP);
  pinMode(openSensorPIN, INPUT_PULLUP);
  pinMode(relayPin, OUTPUT);

  //digitalWrite(relayPin, HIGH);
} // end of setup()

//////////////////////////////////////

void loop(){

  homeSpan.poll();
} // end of loop()

_____________________________________________
File "DEV_DoorsWindows.h"
_____________________________________________
int relayPin = 5;
const int closeSensorPIN = 22;  // GPIO pin for Hall Sensor 1
const int openSensorPIN = 21;  // GPIO pin for Hall Sensor 2

void controlRelay(bool state) { 
  if (state == true) {
    digitalWrite(relayPin, LOW);  // Activate the relay
  } else {
    digitalWrite(relayPin, HIGH); // Deactivate the relay
  }
    // No need to handle deactivation here, as it's handled after the specified duration
}

struct DEV_GarageDoor : Service::GarageDoorOpener {     // A Garage Door Opener

  SpanCharacteristic *current;            // reference to the Current Door State Characteristic (specific to Garage Door Openers)
  SpanCharacteristic *target;             // reference to the Target Door State Characteristic (specific to Garage Door Openers)  
  SpanCharacteristic *obstruction;        // reference to the Obstruction Detected Characteristic (specific to Garage Door Openers)

  DEV_GarageDoor() : Service::GarageDoorOpener(){       // constructor() method
        
    current=new Characteristic::CurrentDoorState(1);              // initial value of 1 means closed
    target=new Characteristic::TargetDoorState(1);                // initial value of 1 means closed
    obstruction=new Characteristic::ObstructionDetected(false);   // initial value of false means NO obstruction is detected
    

    new SpanButton(openSensorPIN);
    new SpanButton(closeSensorPIN);
    
    //this->openSensorPIN=openSensorPIN;                        // NEW! save power pushbutton pin number
    //this->openSensorPIN=openSensorPIN; 

    Serial.print("Configuring Garage Door Opener");   // initialization message
    Serial.print("\n");

  } // end constructor

  boolean update(){                              // update() method

    // see HAP Documentation for details on what each value represents

    if(target->getNewVal()==0){                     // if the target-state value is set to 0, HomeKit is requesting the door to be in open position
      Serial.print("Opening Garage Door\n");
      controlRelay(true);
      current->setVal(2);                           // set the current-state value to 2, which means "opening"
      obstruction->setVal(false);                   // clear any prior obstruction detection
    } else {
      Serial.print("Closing Garage Door\n");                // else the target-state value is set to 1, and HomeKit is requesting the door to be in the closed position
      controlRelay(true);
      current->setVal(3);                           // set the current-state value to 3, which means "closing"         
      obstruction->setVal(false);                   // clear any prior obstruction detection
    }
    
    return(true);                               // return true
  
  
  } // update

void button(int pin) {

    if(pin==openSensorPIN){
      current->setVal(0);
    } else if(pin==closeSensorPIN){                                                   
      current->setVal(1);
    } 
  }

  void loop(){                                     // loop() method

    int ClosedSensorValue = digitalRead(closeSensorPIN); // 0 = magnet lined up with sensor
    int OpenedSensorValue = digitalRead(openSensorPIN); // 1 = no magnet lined up

    if (ClosedSensorValue == 0) {
      target->setVal(1); // value of 1 means the door is closed
      Serial.println("Closed");
      controlRelay(false);
    } else if (OpenedSensorValue == 0) {
      target->setVal(0); // value of 0 means the door is open
      Serial.println("Open");
      controlRelay(true);
    } else {
      target->setVal(1); // if no magnet aligned, the door is open
      Serial.println("Open");
      controlRelay(true);
    }
    
    if(current->getVal()==target->getVal())        // if current-state matches target-state there is nothing do -- exit loop()
      return;

    if(current->getVal()==3 && random(100000)==0){    // here we simulate a random obstruction, but only if the door is closing (not opening)
      current->setVal(4);                             // if our simulated obstruction is triggered, set the curent-state to 4, which means "stopped"
      obstruction->setVal(true);                      // and set obstruction-detected to true
      LOG1("Garage Door Obstruction Detected!\n");
    }

    if(current->getVal()==4)                       // if the current-state is stopped, there is nothing more to do - exit loop()     
      return;

    // This last bit of code only gets called if the door is in a state that represents actively opening or actively closing.
    // If there is an obstruction, the door is "stopped" and won't start again until the HomeKit Controller requests a new open or close action

    if(target->timeVal()>5000)                     // simulate a garage door that takes 5 seconds to operate by monitoring time since target-state was last modified
      current->setVal(target->getVal());           // set the current-state to the target-state
       
  } // loop

};

**Comment by camerong3** [comment]
Still struggling with this if anyone can help

**Reply by danilkorotkov** [reply]
If you want to control motor relays directly from your code you should use interrupts, cycle time control and and predict the current position. 

But the easiest way is to use the built-in motor controller and its buttons open-close-step by step with your relays. 
My gate example with the door

https://github.com/danilkorotkov/HomeSpanGate

I placed optocouplers to connect to internal limit sensors

**Comment by HomeSpan** [comment]
What do you have working so far?  To start, I recommend you eliminate the sensors and only focus on the open/closing that triggers a relay.  To debug, instead of connecting a relay to the ESP32 pin, connect an LED.  This will enable you to check that the pin is turning HIGH or LOW as you expect it.  Once you get all that working and debugged, next step would be to layer in the sensors.

**Reply by prolixmusic** [reply]
I would recommend running a state machine.

The garage door in homekit has 5 states: open, closed, opening, closing and stopped.

'current' is used to provide homekit with the current state of the door. This can also be used for the state of your state machine. Use current->getVal() to get the current state each time you enter the loop. Once the state changes, use current->setVal(*) to change the state. current is what you see on your phone when its says opening, closing, etc. 

'target' is typically set by homekit (from the iPhone app) and tells you where it wants you to go. target represents the position of the door switch on your phone. Use target->getval to get the target state. Use target->timeVal() to see how much time has passed since the last target state change. I use this time to prevent the motor from running indefinitely in the event the motor is uncoupled from the door. I also use it to turn off my light.

target->setVal() should not used unless you have a hardwired door button. I do, as such, if the door is not running, a press of the button, will toggle the target state, which is then read by the state machine to start the door running.

Anyway, in each of the states, you have a specific set of things you need to do to make the door function properly. 

In your sketch, you are using target->setVal() when you should be using current->setVal()...
